jerry-core/api/jerry-module.c

```diff
@@ -0,0 +1,249 @@
+typedef struct jerry_module_t
+{
+  struct jerry_module_t *next_p; /**< next_module */
+  jerry_char_t *path_p; /**< path to the module */
+  jerry_size_t basename_offset; /**< offset of the basename in the module path*/
+  jerry_value_t realm; /**< the realm of the module */
+  jerry_value_t module; /**< the module itself */
+} jerry_module_t;
+
+/**
+ * Native info descriptor for modules.
+ */
+static const jerry_object_native_info_t jerry_module_native_info JERRY_ATTR_CONST_DATA = {
+  .free_cb = NULL,
+};
+
+/**
+ * Default module manager.
+ */
+typedef struct
+{
+  jerry_module_t *module_head_p; /**< first module */
+} jerry_module_manager_t;
+
+/**
+ * Release known modules.
+ */
+static void
+jerry_module_free (jerry_module_manager_t *manager_p, /**< module manager */
+                   const jerry_value_t realm) /**< if this argument is object, release only those modules,
+                                               *   which realm value is equal to this argument. */
+{
+  jerry_module_t *module_p = manager_p->module_head_p;
+
+  bool release_all = !jerry_value_is_object (realm);
+
+  jerry_module_t *prev_p = NULL;
+
+  while (module_p != NULL)
+  {
+    jerry_module_t *next_p = module_p->next_p;
+
+    if (release_all || module_p->realm == realm)
+    {
+      jerry_port_path_free (module_p->path_p);
+      jerry_value_free (module_p->realm);
+      jerry_value_free (module_p->module);
+
+      jerry_heap_free (module_p, sizeof (jerry_module_t));
+
+      if (prev_p == NULL)
+      {
+        manager_p->module_head_p = next_p;
+      }
+      else
+      {
+        prev_p->next_p = next_p;
+      }
+    }
+    else
+    {
+      prev_p = module_p;
+    }
+
+    module_p = next_p;
+  }
+} /* jerry_module_free */
+
+/**
+ * Initialize the default module manager.
+ */
+static void
+jerry_module_manager_init (void *user_data_p)
+{
+  ((jerry_module_manager_t *) user_data_p)->module_head_p = NULL;
+} /* jerry_module_manager_init */
+
+/**
+ * Deinitialize the default module manager.
+ */
+static void
+jerry_module_manager_deinit (void *user_data_p) /**< context pointer to deinitialize */
+{
+  jerry_value_t undef = jerry_undefined ();
+  jerry_module_free ((jerry_module_manager_t *) user_data_p, undef);
+  jerry_value_free (undef);
+} /* jerry_module_manager_deinit */
+
+/**
+ * Declare the context data manager for modules.
+ */
+static const jerry_context_data_manager_t jerry_module_manager JERRY_ATTR_CONST_DATA = {
+  .init_cb = jerry_module_manager_init,
+  .deinit_cb = jerry_module_manager_deinit,
+  .bytes_needed = sizeof (jerry_module_manager_t)
+};
+
+#endif /* JERRY_MODULE_SYSTEM */
+
+/**
+ * Default module resolver.
+ *
+ * @return a module object if resolving is successful, an error otherwise
+ */
+jerry_value_t
+jerry_module_resolve (const jerry_value_t specifier, /**< module specifier string */
+                      const jerry_value_t referrer, /**< parent module */
+                      void *user_p) /**< user data */
+{
```

Comment:
What's the intension of user_p and not used by this function.

Reply:
It is there to match the interface required by jerry_module_link. 

Normally, jerry_module_resolve is not called directly, instead it is a callback in jerry_module_link. Embedders can also create custom module resolvers, and make jerry_module_link use those as callbacks.

The user_p is passed by jerry_module_link, which allows user data to be passed to the callback function in an arbitrary way. The default jerry_module_resolve does not require any extra user data, thats why it's unused, however another implementaion may do so.